File: src/test/java/quebec/virtualite/backend/utils/RestClientTest.java
package quebec.virtualite.backend.utils;

import lombok.val;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static quebec.virtualite.backend.utils.RestParam.param;

@ExtendWith(MockitoExtension.class)
class RestClientTest
{
    @InjectMocks
    private RestClient client;

    @Test
    void urlWithParams()
    {
        urlWithParams(
            "/a/b={b}", List.of(param("b", 1)),
            "/a/b=1");

        urlWithParams(
            "/a", List.of(param("b", 1)),
            "/a?b=1");

        urlWithParams(
            "/a", List.of(param("b", 1), param("c", 2)),
            "/a?b=1&c=2");

        urlWithParams(
            "/a/b={b}", List.of(param("b", 1), param("c", 2), param("d", 3), param("e", 4)),
            "/a/b=1?c=2&d=3&e=4");
    }

    private void urlWithParams(String urlStart, List<RestParam> params, String urlResult)
    {
        // When
        val url = client.urlWithParams(urlStart, params.toArray(new RestParam[0]));

        // Then
        assertThat(url).isEqualTo(urlResult);
    }
}

File: src/test/java/quebec/virtualite/backend/utils/RestClientTest.kt
package quebec.virtualite.backend.utils

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
@Tag("UnitTest")
class RestClientTest
{
    @InjectMocks
    private lateinit var client: RestClient

    @Test
    fun urlWithParams()
    {
        urlWithParams(
            "/a/b={b}",
            arrayOf(RestParam("b", 1)),
            "/a/b=1"
        )

        urlWithParams(
            "/a",
            arrayOf(RestParam("b", 1)),
            "/a?b=1"
        )

        urlWithParams(
            "/a",
            arrayOf(RestParam("b", 1), RestParam("c", 2)),
            "/a?b=1&c=2"
        )

        urlWithParams(
            "/a/b={b}",
            arrayOf(RestParam("b", 1), RestParam("c", 2), RestParam("d", 3), RestParam("e", 4)),
            "/a/b=1?c=2&d=3&e=4"
        )
    }

    private fun urlWithParams(urlStart: String, params: Array<RestParam>, urlResult: String)
    {
        // When
        val url = client.urlWithParams(urlStart, params)

        // Then
        assertThat(url).isEqualTo(urlResult)
    }
}
